lib/rules/selector-max-type/__tests__/index.mjs

```diff
@@ -109,25 +106,25 @@ testRule({
 			line: 1,
 			column: 13,
 			endLine: 1,
-			endColumn: 20,
+			endColumn: 21,
```

Comment:
[ask] `13` to `20` seems correct to me 👀 

```css
.foo, .bar, foo.baz {}
/*          ↑      ↑
12345678901234567890
*/
```

Reply:
Ah, I see. The selector is trimmed, right?

Reply:
Thank you for calling this out.

To be honest I am never sure about these and I am prone to having off by one errors here.

Reply:
I had an issue here, but I want to take a closer look again later.

I think the end index needs to be one less than it currently is.
The doc comment for `endIndex` also states that it is an exclusive index, so that would make sense.

But for `a {}` it is then `[0, 0]` and this causes other issues.
The final `endColumn` then becomes `5`.

Not sure if this makes sense and if someone knows what issue I've encountered? :)

Reply:
I'm sorry, I didn't understand your question well. 😓 

The `a {}` CSS code has the location `1:1-1:5` when parsed by PostCSS, right?

```sh-session
$ node -pe 'require("postcss").parse("a {}").source'
{
  input: Input {
    css: 'a {}',
    hasBOM: false,
    id: '<input css CWnXw->',
    [Symbol(fromOffsetCache)]: [ 0 ]
  },
  start: { column: 1, line: 1, offset: 0 },
  end: { column: 5, line: 1, offset: 4 }
}
```

Reply:
This is a bug in both PostCSS and Stylelint as far as I can tell.

see:  https://github.com/postcss/postcss/blob/0fd1d86f2d33926c782463e63d8e1c816d207bc8/lib/node.js#L244-L248

see: https://github.com/stylelint/stylelint/blob/a02e4dd6d99e9159fd0f51d04caf324c1d990066/lib/utils/report.mjs#L111

If `endIndex` is `0` it is treated as if it isn't set, which isn't correct.

It is just really really rare for the `endIndex` to be `0` in CSS.
Only happens for a type selector with only a single `char`.

So users would have to start their CSS file with `a {}`, `i {}`, `b {}`, ... or something similar.

Reply:
[suggest] Thanks for the fix. However, it seems that we still need to fix the calculation of `endIndex` when trimming whitespaces in a selector string.

- `.length` has to be calculated before trimming
- `endIndex` has to be without `-1` (excluding)

```js
const selectorStr = selectorNode.toString().trim();
const index = selectorNode.first?.sourceIndex ?? 0;
const endIndex = index + selectorStr.length;

report({
	ruleName,
	result,
	node: ruleNode,
	message: messages.expected,
	messageArgs: [selectorStr, primary],
	index,
	endIndex,
});
```

```css
/* start: 31, end: 42 */
.foo + .bar { & .baz + .foo { & foo + bar {} } }
                              ^          ^
123456789012345678901234567890123456789012
```

Any thoughts?

Reply:
🤔 I am not sure I follow.

If I test these I get the ranges I would expect:

<img width="971" alt="Screenshot 2024-03-21 at 19 34 26" src="https://github.com/stylelint/stylelint/assets/11521496/ce8bd1c0-a9c8-415f-af44-35f3bc83110f">

The selector text is also only trimmed for the message. The length value is from the untrimmed selector.

But happy to make any needed adjusted, just trying to understand it better :)

Reply:
[`stylelint-codeframe-formatter`](https://www.npmjs.com/package/stylelint-codeframe-formatter) helps us see the reported position.

```jsonc
// .stylelintrc.json
{
  "rules": {
    "selector-max-type": 0
  }
}
```

```css
/* test.css */
.foo + .bar { & .baz + .foo { & foo + bar {} } }
```

Run:

```sh-session
$ npx stylelint test.css --custom-formatter node_modules/stylelint-codeframe-formatter/index.js
  test.css

  ✖ Expected "& foo + bar" to have no more than 0 type selectors (selector-max-type) at test.css:2:31
  1 | /* test.css */
> 2 | .foo + .bar { & .baz + .foo { & foo + bar {} } }
    |                               ^^^^^^^^^^
  3 |

  1 error
```

As the output above shows, the end position is short by one character.

Reply:
It isn't fixed yet with my last commit, but I fully understand the issue now.
Will look at it later today, or early tomorrow.

Thank you for the extra information here 🙇

Reply:
`sourceIndex` on `postcss-selector-parser` AST nodes doesn't include the length of `rawSpaceBefore`.

Any selector with a comma followed by a space (e.g. `foo, bar`) had unexpected results. `sourceIndex` on `bar` would be without the leading space. While the string representation would include the leading space.

This means that `sourceIndex` + `length` wouldn't be correct when used as `endIndex`.

Trimming all surrounding whitespace first and then taking the `length` seems to fix it.

@ybiquitous Please let me know if this makes sense :)

Reply:
Thanks a lot! The changes make sense!